diff --git a/cantor/src/backends/CMakeLists.txt b/cantor/src/backends/CMakeLists.txt
index 80a8651..9706258 100644
--- CMakeLists.txt
+++ CMakeLists.txt
@@ -14,3 +14,9 @@ macro_log_feature(R_FOUND "R" "A free software environment for statistical compu
 if(R_FOUND)
 	macro_optional_add_subdirectory(R)
 endif(R_FOUND)
+
+macro_optional_find_package(Qalculate)
+macro_log_feature(QALCULATE_FOUND "Qalculate" "A multi-purpose desktop calculator with support for customizable functions, units and arbitrary precision." "http://qalculate.sourceforge.net/" FALSE "" "Backend to use Qalculate with Cantor.")
+if(QALCULATE_FOUND)
+  macro_optional_add_subdirectory(qalculate)
+endif(QALCULATE_FOUND)
